### PR TITLE
Measured field maps

### PR DESCRIPTION
**Pull request by jonpaley**
Overhaul of the MagneticField package.  The biggest changes are:

- the field maps are now in our /cvmfs directory and read from there
- the magnetic field is now positioned correctly
- testing functionality has been moved to a test subdirectory
- dependence on Geant4 has been removed

The critical functionality of the field class (loading from a text file, accessing the field via either a STL vector or map (vector is the default), calculating an integral and finally the "GetFieldValue" method needed by Geant4 remains the same.

Along the way I have updated the event display to optionally draw true particle trajectories (needs some tuning so that we're not overwhelmed with low-energy electrons) and the MCValidation package to create TGraphs of sim::Particles for validation.  From my tests, besides seeing particles bending at a different (but I think correct) location, things look pretty consistent with what we currently simulate in the "main" branch.

### CODE REVIEW

**Review by gavinsdavies — APPROVED** [review]
A good amount of changes here. At a glance looks good. If more detailed review is needed let me know but given feedback on tests, it sounds like it's doing as advertised

**Comment by paullglebrun** [comment]
I presumed I don't have have to do anything in git, as Gavin already approved the pull... Rightfully so.. Let me know if this is not the case. ..

Jon, you made a lot of progress!

My tentative plan:  (i) After removing my infamous "fIsMC" flag in my SSD reconstruction code, I probably should spend a bit more time cleaning things up and sorted out is the stereo angle (45 degrees, or 45 + 90 degrees, for U and W, respectively, or the other way around, for station 2 through 5. ).  With real data., run 1274, no target but a target, at 31 GeV. Check again the MC.  (ii) re-aling 1274, as the  Z positions changed a bit, and hopefully trying to include some beam constrains, perhaps.. (iii) Merge main into my stuff, such that I can start using the measuredField.

There is one more thing I briefly discussed with Linyan, that is, on the current main branch, how do we (i) tweak the transverse position of the magnet, in MC and reconstruction? (ii) Related to this,  the Geometry class does not give us the transverse position of the magnet center.. Only the Z position.

Cheers,

Paul

________________________________

**Comment by jonpaley** [comment]
Hi Paul,

	Right, I wanted to make sure that you were aware of the changes.  I don’t think I removed anything that is critical to your alignment studies, but we need to keep a careful eye on things since my validation studies were not completely exhaustive.  You will want to merge “main” into your feature branch after I do this merge and let us know if you run into any problems.

	For tweaking the magnet position, my approach to this (for systematic studies) is the same as one would do for any other alignment study: simulate with a single a single Geometry, and reconstruct with a “different Geometry”.  By “different”, I mean that we apply a translation and/or rotation to (in this case) the magnet (and field).  We’ll have to think carefully about how to do this in a consistent way, so that both the magnet material and the field itself undergo the same transformation.  But we also need to be sure we don’t double count… a transformation on the magnet but keeping the SSDs fixed is the same as applying a transformation to the SSDs but keeping the magnet fixed.  So if we have an uncertainty on the position and orientation of the magnet, we can study the affect by shifting the positions and orientations of the SSDs.  Right now I’m thinking it might be easier to do this to the SSDs, but regardless, some thought needs to go into how we want to put the hooks into the Geometry for this.  Given a Root Geometry, this is all very straightforward, and the reconstruction will be using the Root Geometry.

	I will hold off for ~12 hours, and pull the trigger on this PR tomorrow morning.  Not a good idea to make these changes at the end of the day!

	Best,

	Jon


Jonathan M. Paley (he/him/his)
Scientist

Neutrino Division
Fermi National Accelerator Laboratory
630 840 5290 office
www.fnal.gov <http://www.fnal.gov/>

Connect with us!
Newsletter <http://www.fnal.gov/pub/today/>  |  Facebook <https://www.facebook.com/Fermilab>  |  Twitter <https://twitter.com/FermilabToday>

**Pull request merged**
